Integrate TACo and OrbisDB

theref opened the pull request:
These changes aim to integrate TACo encryption/decryption to OrbisDB DeForum demo.

Posts are encrypted and a decryption condition is set: balance in native token must be positive.

Posts are decrypted if condition is met.

**Required reviews:** 2

**Note for reviewers:**
For some reason GitHub is not redering well these changes, probably because I did a rebase. To check the actual differences in the code: https://github.com/nucypher/taco-orbisdb-demo/compare/main...taco

theref committed:
Begin adding taco

theref committed:
Initialize the wasm

theref committed:
Use strings when posting encrypted message

theref committed:
Fix uploading of encrypted message

theref committed:
Get encryption and decryption working

theref committed:
Reuse signature for decryption

theref committed:
Address PR review requests

theref committed:
Update README

manumonti committed:
Add TACo description to README

theref committed:
Fix README and improve calling of initialize

theref committed:
Update README.md

Co-authored-by: Derek Pierre <derek.pierre@gmail.com>

theref committed:
Refactor handling of AuthProvider

manumonti commented on README.md:
````diff
@@ -1,22 +1,21 @@
 
 ```bash
-git clone https://github.com/ceramicstudio/deforum && cd deforum
+git clone https://github.com/nucypher/orbis-demo && cd orbis-demo
````
```suggestion
git clone https://github.com/nucypher/taco-orbisdb-demo && cd taco-orbisdb-demo
```

manumonti commented on app/taco.ts:
```diff
@@ -0,0 +1,97 @@
+    currentAccount: null,
+
+    async initialize() {
+      if (!this.provider) {
+        this.provider = new ethers.providers.Web3Provider(window.ethereum);
+        await this.updateAccount();
+        this.setupAccountChangeListener();
+      }
+      return this.authProvider;
```
When I run the demo, I get the following error:

```
  10 |     async initialize() {
  11 |       if (!this.provider) {
> 12 |         this.provider = new ethers.providers.Web3Provider(window.ethereum);
     |                                                           ^
  13 |         await this.updateAccount();
  14 |         this.setupAccountChangeListener();
  15 |       }
 ⨯ app/taco.ts (12:59) @ window
 ⨯ unhandledRejection: ReferenceError: window is not defined
    at Object.initialize (webpack-internal:///(ssr)/./app/taco.ts:21:82)
    at getAuthProvider (webpack-internal:///(ssr)/./app/taco.ts:42:24)
    at Posts (webpack-internal:///(ssr)/./components/sections/posts.tsx:35:74)
    at au (/Users/manumonti/Projects/nucypher/taco-orbisdb-demo/node_modules/.pnpm/next@14.2.11_@babel+core@7.25.2_react-dom@18.3.1_react@18.3.1/node_modules/next/dist/compiled/next-server/app-page.runtime.dev.js:35:10446)
```

My guess is that Next.js renders the components on the browser, but also in the server. In the server there is no `window` object, so this error is raised on the terminal: https://www.geeksforgeeks.org/how-to-fix-the-error-window-is-not-defined-in-nextjs/

```suggestion
      if(typeof window !== 'undefined') {
        if (!this.provider) {
          this.provider = new ethers.providers.Web3Provider(window.ethereum);
          await this.updateAccount();
          this.setupAccountChangeListener();
        }
        return this.authProvider;
      }
```

derekpierre replied:
Analogous in `taco-composedb` - https://github.com/nucypher/taco-composedb/blob/main/utils/index.ts#L19.

derekpierre replied:
Be careful because initialize is only called once, so if `window.ethereum` is not defined at that time, you will never have a `this.provider` that is set.

theref replied:
@manumonti how did you run the app? i haven't seen this error at all

theref replied:
i added a retry mechanism for `window.ethereum`

manumonti replied:
> @manumonti how did you run the app? i haven't seen this error at all

With `pnpm dev`. Maybe is something related to the node version that is being used? Here are the full logs:

https://gist.github.com/manumonti/d00bdd177bda546927d25c2c281fc320

derekpierre commented on app/taco.ts:
```diff
@@ -0,0 +1,97 @@
+
+    async initialize() {
+      if (!this.provider) {
+        this.provider = new ethers.providers.Web3Provider(window.ethereum);
```
Just to confirm, so we never need to re-instantiate a provider when the account changes? It's internals get automatically updated when the account in metamask etc. is switched...? (Pardon my ignorance of ethers)

theref replied:
That's certainly my understanding of how ethers works

manumonti reviewed (changes requested):
I see a little bug: when you click on a Thread, the body of the post is not decrypted. I'm coding a fix for this.

![Screenshot 2024-10-02 at 17 03 23](https://github.com/user-attachments/assets/be7c6a77-9ea1-4933-bb26-bc5cc9279162)

Edit: PR https://github.com/nucypher/taco-orbisdb-demo/pull/2

manumonti committed:
Fix the thread body not being decrypted

theref committed:
Update README.md

Co-authored-by: Manuel Montenegro <manuel@nucypher.com>

mzkrasner committed:
Removing the need for NextAuth  secret

derekpierre reviewed (dismissed):
@KPrasch , @manumonti and I had a joint session for @manumonti to show us how the demo works. We are still prompting for signatures for each post.

As part of that session, we discovered that the demo actually uses SIWE for user management—https://github.com/nucypher/taco-orbisdb-demo/blob/main/app/context/OrbisContext.tsx#L83. As such, we should use the single sign-on auth provider instead and the existing SIWE signature utilized from the `OrbisSession`. This is analogous to what was done in th `taco-composedb` demo for `DIDSession` i.e.
- create a utility to get the siwe message and signature from the underlying session - https://github.com/nucypher/taco-composedb/blob/main/utils/index.ts#L76
- use the utility to get the info and create a single sign-on auth provider - https://github.com/nucypher/taco-composedb/blob/main/src/fragments/chatcontent.tsx#L43-L44

This will resolve the prompting for signatures.

manumonti committed:
Apply prettier format rules

manumonti committed:
Add prettier formatting to package.json scripts

manumonti committed:
Modify TACo utilities to be a react hook

manumonti committed:
Modify useTaco hook to use context variables

manumonti committed:
Adapt encryption/decryption calls to new TACo hook

manumonti committed:
Add check for user instance

manumonti committed:
Remove some logs

mzkrasner committed:
Upgrading orbis sdk version with changes to authentication pattern

mzkrasner committed:
Reverting readme instructions back to using pnpm

manumonti committed:
Regenerate pnpm-lock after rebasing main

manumonti committed:
Add TACo context to store global variables

manumonti committed:
Add didtools package

manumonti committed:
Reuse orbis session SIWE for TACo decryption

manumonti committed:
Format code with prettier

manumonti committed:
Return <Decryption failed> string when it fails

manumonti committed:
Add enum to Zod check of TACo domain in .env file

manumonti committed:
Merge pull request #4 from manumonti/siwe-reuse

Implement the reuse of SIWE messages from OrbisDB session

derekpierre commented on app/(home)/(posts)/post/[slug]/page.tsx:
```diff
@@ -83,41 +85,55 @@ export default function PostPage({
   const getPost = async (stream_id: string): Promise<void> => {
+    if (!window.ethereum) {
+      console.error("No Ethereum provider found");
+      return;
```
Is there a difference between using `return;` (here) and `return undefined;` (L140)?

This same scenario happens in other areas of the demo.

manumonti replied:
> Is there a difference between using `return;` (here) and `return undefined;` (L140)?
> 
> This same scenario happens in other areas of the demo.

Actually `return;` and `return undefined;` is the same.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/return

I prefer just `return` because I find it more aesthetic, but the alternative is ok as well.

https://github.com/nucypher/taco-orbisdb-demo/pull/6

derekpierre commented on app/hooks/useTaco.ts:
```diff
@@ -0,0 +1,101 @@
+    const siweInfo = await loadSiweFromOrbisSession();
+    if (!siweInfo.message || !siweInfo.signature) {
+      console.error("SIWE message or signature missing");
+      return "<Decryption failed>";
```
Perhaps something to indicate a different failure mode?

```suggestion
      return "<User not already logged in>";
```

manumonti replied:
It's a more comprehensible string indeed. The problem that I see is that, if I'm understanding well, the lack of `siweInfo.message` or `siweInfo.signature` can come from an error when the [async function `DIDSession.fromSession(session)` is called](https://github.com/nucypher/taco-orbisdb-demo/blob/ee218bc87b623a17dd97c2ce03cce1d06556ee2f/app/hooks/useTaco.ts#L85-L86). 

This function eventually will call this function on `dids` package and it seems that there are multiple causes that can lead an error that is not related with the "User not already logged in": https://github.com/ceramicnetwork/js-did/blob/def0a5fed152a42ffeef87bafb4619816f9954f6/packages/dids/src/did.ts#L248

So I would prefer to have a more generic error. Does it make sense to you?

arjunhassard replied:
```suggestion
      return "<Decryption failed. This may be because the user has not yet logged in.>";
```

manumonti committed:
Remove Sepolia from networks that can be selected

When trying TACo encryption or decryption with Sepolia network selected,
it returns an error: "Error: No contract found for name: Coordinator".

If you select Sepolia, the web3 provider (Metamask) will try to call the
Coordinator contract on Sepolia, but the testnet where this contract
lives is Polygon Amoy, so this make the app to fail.

manumonti committed:
Standardize the use of return undefined

manumonti committed:
Merge pull request #5 from manumonti/sepolia-removed

Remove Sepolia from networks that can be selected

manumonti committed:
Merge pull request #6 from manumonti/return-undefined

Standardize the use of return undefined

arjunhassard commented on README.md:
```diff
@@ -1,6 +1,6 @@
 # DeForum - A Web Forum on OrbisDB
 
-A web forum demo application built on OrbisDB.
+A web forum demo application built on OrbisDB and utilising TACo for encrypting posts and managing who has access to them.
```
```suggestion
A web forum demo application built on OrbisDB, leveraging TACo to create non-public posts with pre-specified conditions for access. 
```

arjunhassard commented on README.md:
```diff
@@ -9,19 +9,21 @@ A web forum demo application built on OrbisDB.
 This application emulates a web forum, complete with posts, comments, images, and personal profiles. It relies on the following items:
 
 **OrbisDB on Ceramic** - A decentralized database to hold the various forum artifacts users create (such as posts, comments, etc.)
+**TACo** - A protocol for secure, end-to-end encrypted data sharing and communication, enabling encryption and decryption of data based on access control logic.
```
```suggestion
**TACo** - A decentralized access control layer, through which decryption rights to non-public forum content (posts, comments, reactions, etc.) are verified and enforced by a decentralized cohort of TACo nodes. 
```

arjunhassard reviewed (approved):
Let's merge this, then solicit feedback and further design/improvement directions from the OrbisDB team

manumonti committed:
Update README.md with better wording

Co-authored-by: Arjun Hassard <arjunhassard@gmail.com>

manumonti committed:
Update README with better wording

Co-authored-by: Arjun Hassard <arjunhassard@gmail.com>

manumonti committed:
Add a more insightful error message

Co-authored-by: Arjun Hassard <arjunhassard@gmail.com>
